Change S rank to require no miss

peppy opened the pull request:
As proposed in the [original proposal](https://github.com/ppy/osu/issues/4184#issuecomment-1724954327). Got left out somewhere along the way.

- [x] Fix ranking screen display to handle this better
- [x] Migration considerations
- [x] More testing, probably

---

https://github.com/ppy/osu/assets/191335/efff388d-1a3f-45ad-9f2a-6966a999f4f9

bdach committed:
Adjust existing test to fail

bdach committed:
Move maximum statistics population to `LegacyScoreDecoder`

bdach committed:
Use lazer accuracy & rank implementations across the board

peppy committed:
Change S rank to require no miss

peppy committed:
Update test in line with new rank definitions always being applied

peppy committed:
Move `BackgroundDataStoreProcessor` to better namespace

peppy committed:
Add migration

peppy committed:
Adjust results screen to handle S->A rank adjustment when misses are present

peppy commented:
This is in a pretty good state now, except migrations which will lean on @bdach's work with https://github.com/ppy/osu/issues/26595.

bdach committed:
Merge branch 'master' into s-rank-change

bdach committed:
Make `RankFromScore()`'s dictionary param readonly

Just to make sure nobody tries any "funny" business.

bdach committed:
Fix back-to-front conditional in taiko processor

Would be weird to degrade a score due to *no* misses wouldn't it?

bdach committed:
Remove remnant of old implementation of showing "A due to misses"

bdach committed:
Add another failing tests covering recomputation of ranks

bdach committed:
Unify legacy total score / accuracy / rank recomputation flows

bdach committed:
Remove superfluous recomputation of accuracy

bdach committed:
Remove legacy computation of accuracy & ranks

bdach committed:
Update assertions to match expected behaviour

bdach committed:
Expand xmldoc

bdach committed:
Fix rank background processing not working (and not bumping score version)

bdach commented:
I've just pushed my migration changes, but also some fixes to the previous code.

Of note are: https://github.com/ppy/osu/pull/26630/commits/cb8ec48717b49110bb784ae69028612d546a7a40 (some extra API safety), https://github.com/ppy/osu/pull/26630/commits/20ed7e13e31c4da412e1cffab581e4b7054492cb (bug), https://github.com/ppy/osu/pull/26630/commits/fdd499a4859bcd682dc09156b3dcef1a8a32bcb8 (bug).

Rest is migration code & general cleanup that was permitted by it. It has been lightly tested but not too much. I will do more testing of it tomorrow if I don't wake up to this being merged.

bdach committed:
Remove redundant assertion

It does look pretty dumb to be checking if something is `null` *after
calling an instance method on it*...

peppy commented:
20ed7e1 is weird because i could've sworn i fixed and force pushed this very early...

juliand665 commented:
I'm a little confused about this change, particularly this rationale:

> Got left out somewhere along the way.

I just spent a while trying to find any place where people were formally polled about this, and I suppose the linked comment is basically that, however, the discussion did not end there. https://github.com/ppy/osu/issues/4184#issuecomment-1727415278 saw near-unanimous approval for what I interpreted as basing grades solely on accuracy (since FC is already represented in other ways). Precisely this seems to be what was ultimately moved forward with (https://github.com/ppy/osu/issues/4184#issuecomment-1742503532), so I'm not sure where this reversal is coming from.

peppy commented:
> so I'm not sure where this reversal is coming from.

User acceptance is the issue. We're staying with what people are used to (from stable) so we *can make forward progress*. Thanks for understanding.

juliand665 commented:
Thanks for clarifying! I wish I could see a formal vote on this or something like that, but I assume this kind of insight is just gleaned from the nebulous general consensus across different social media? That's unfortunate, but I guess I can live with it.

peppy commented:
Our goal right now is to get things live and start to get feedback from actually playing the game against existing leaderboards.

This change is easier than trying to figure some "FC" marker that people are happy with.

Grades can be recalculated at any point, so if there's enough community push we can change them down the road based on more well formed feedback.

juliand665 commented:
That makes sense, thank you for explaining! Can't wait to see lazer become the new face of osu :)

bdach commented:
Provisional osu! score spreadsheet with these changes: https://docs.google.com/spreadsheets/d/16FPqye6GWHZDZPAd0zPsRd3JTBxeNdjRF7LiP1EjgAI/edit#gid=364418891

I'm not sure why scores are getting nerfed to the ground. Will require an investigation (on top of fixing the failing tests).

Walavouchey commented on osu.Game/Screens/Ranking/Expanded/Accuracy/AccuracyCircle.cs:
```diff
@@ -387,6 +397,31 @@ protected override void LoadComplete()
+                                .FadeOut(800, Easing.Out);
+
+                            accuracyCircle
+                                .FillTo(accuracyS - 0.0001, 70, Easing.OutQuint);
```
some special consideration was added in https://github.com/ppy/osu/pull/22554/files#diff-6ad7f21c9f785ced34f1b5dea121ccf10e8ff29baae1b2ec9b08a30b4bae6cd3R289-R312 in order to prevent the circle fill from landing in the wrong sector or in the gaps between them

this branch (it looks like it landed on S):

![image](https://github.com/ppy/osu/assets/36758269/e95287ec-8159-4ed6-95a0-84d1b9af0f4d)

proposed fix:

![image](https://github.com/ppy/osu/assets/36758269/27f5af86-f912-4177-8b05-a58550c54d30)

```suggestion
                                .FillTo(accuracyS - NOTCH_WIDTH_PERCENTAGE / 2 - visual_alignment_offset, 70, Easing.OutQuint);
```

bdach committed:
Move transferal of `LegacyTotalScore` back to original spot

This subtle detail was messing with server-side score import flows.
Server-side, legacy total score will *already* be in `LegacyTotalScore`
from the start, and `TotalScore` will be zero until recomputed via
`StandardisedScoreMigrationTools.UpdateFromLegacy()` - so in that
context, attempting to move it across is incorrect.

peppy committed:
Apply more correct visual offset adjustment

Co-authored-by: Walavouchey <36758269+Walavouchey@users.noreply.github.com>

bdach committed:
Remove outdated test

Non-legacy scores *are* subject to upgrades again - albeit it is *rank*
upgrades that they are subject to.

bdach committed:
Do not reprocess ranks for custom rulesets

Chances are that because we've broken rank API, it would utterly fail
for all custom rulesets anyhow.

bdach commented:
The sheet breakage mentioned above is fixed by https://github.com/ppy/osu/pull/26630/commits/cb87d6ce500bd8070b1a0ad69537a6f643e1aa67. Tested locally using the server-side verify command.

I think my migration portion is ready for light review at least, although some further testing is probably warranted. To that end I'll queue up a score calc run now and hopefully it passes given https://github.com/smoogipoo/diffcalc-sheet-generator/pull/1 is now merged...

If it passes then I'll queue up the rest just to be sure. osu! should be matching as per above so I'll skip it.

!diffcalc
GENERATORS=score
RULESET=taiko
SCORE_PROCESSOR_A=https://github.com/bdach/osu-queue-score-statistics/tree/fix-compile-failure
SCORE_PROCESSOR_B=https://github.com/ppy/osu-queue-score-statistics/pull/203

bdach commented:
Empty is a good omen...

!diffcalc
GENERATORS=score
RULESET=catch
SCORE_PROCESSOR_A=https://github.com/bdach/osu-queue-score-statistics/tree/fix-compile-failure
SCORE_PROCESSOR_B=https://github.com/ppy/osu-queue-score-statistics/pull/203

bdach commented:
!diffcalc
GENERATORS=score
RULESET=mania
SCORE_PROCESSOR_A=https://github.com/bdach/osu-queue-score-statistics/tree/fix-compile-failure
SCORE_PROCESSOR_B=https://github.com/ppy/osu-queue-score-statistics/pull/203

bdach commented:
Sheets look good. Not sure there is much more testing I can do here.

peppy commented on osu.Game/Database/BackgroundDataStoreProcessor.cs:
```diff
@@ -376,6 +376,66 @@ private void convertLegacyTotalScoreToStandardised()
+                 .AsEnumerable()
+                 // must be done after materialisation, as realm doesn't support
+                 // filtering on nested property predicates or projection via `.Select()`
+                 .Where(s => s.Ruleset.IsLegacyRuleset())
```
Right, I removed this thinking it was checking legacy scores, but turns out this is important I guess.

peppy commented:
https://github.com/ppy/osu/assets/191335/da0da67c-9654-4cbf-979d-181a637ed957
